### PR TITLE
Config override and system properties for tests cleanup

### PR DESCRIPTION
**Pull request by randomanderson**
_My *third* attempt at fixing config overrides for tests.  Previous failures: #1034, #1007_

### The goals of this PR are:

- Make it impossible for system config or environment variables to leak between tests
- Eliminate the need to manually cleanup any properties
- Simplify the "Config override" interface
- Eliminate `static` code blocks in tests that alter config or set properties

### To that end, the new interface is:
- `injectSysConfig(name, value)`
- `removeSysConfig(name)`
- `injectEnvConfig(name, value)`
- `removeEnvConfig(name)`

All methods update the `Config` instance to reflect the current state.

### Tests extending `AgentTestRunner`
The main use case for static blocks were tests that needed to set properties before the agent instrumentation runs.  These tests can now override `configurePreAgent()` and call any of the above methods.

### Future work
- There are tests that can't extend `DDSpecification`.  They still need to manually clear config.  This will be addressed in a followup PR.
- The tests reveal the RabbitMQ instrumentation is broken w.r.t. e2e duration.  This was hidden by the fact that the config was always a certain way

### CODE REVIEW

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/base/HttpServerTest.groovy [hunk 1]
```diff
@@ -309,12 +309,13 @@ abstract class HttpServerTest<SERVER> extends AgentTestRunner {
...
     def request = request(endpoint, method, body).build()
-    Response response = withConfigOverride("http.server.tag.query-string", "true") {
-      client.newCall(request).execute()
-    }
 
-    expect:
+    when:
+    Response response = client.newCall(request).execute()
+
+    then:
```
Comment (tylerbenson) [inline dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/base/HttpServerTest.groovy:318]: Is changing expect -> when/then an important/required change or just done for consistency/style? (I'm fine with it, just making sure I understand.)

File: dd-trace-core/src/test/groovy/datadog/trace/core/util/SystemAccessTest.groovy [hunk 2]
```diff
@@ -106,10 +102,8 @@ class SystemAccessTest extends DDSpecification {
 
   def "Test getVMArguments"() {
     setup:
-    ConfigUtils.updateConfig {
-      System.setProperty(PREFIX + PROFILING_ENABLED, profilingEnabled.toString())
-      System.setProperty(PREFIX + HEALTH_METRICS_ENABLED, healthMetricsEnabled.toString())
-    }
+    injectSysConfig(PROFILING_ENABLED, profilingEnabled.toString())
+    injectSysConfig(HEALTH_METRICS_ENABLED, healthMetricsEnabled.toString())
```
Comment (tylerbenson) [inline dd-trace-core/src/test/groovy/datadog/trace/core/util/SystemAccessTest.groovy:96]: Probably not a big deal, but the result of this change is that config is reconstructed twice instead of just once.

File: utils/test-utils/src/main/groovy/datadog/trace/test/util/DDSpecification.groovy [hunk 3]
```diff
@@ -15,13 +20,24 @@ import static net.bytebuddy.matcher.ElementMatchers.none
...
   // Keep track of config instance already made modifiable
   private static isConfigInstanceModifiable = false
 
+  @Rule
+  public final ResetControllableEnvironmentVariables environmentVariables = new ResetControllableEnvironmentVariables()
+
+  // Intentionally not using the RestoreSystemProperties @Rule because this needs to save properties
+  // in the BeforeClass stage instead of Before stage.  Even manually calling before()/after
+  // doesn't work because the properties object is not cloned for each invocation
+  private static Properties originalSystemProperties
```
Comment (tylerbenson) [inline utils/test-utils/src/main/groovy/datadog/trace/test/util/DDSpecification.groovy:39]: 👍

File: dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/test_before_1_11_106/groovy/AWS0ClientTest.groovy [hunk 4]
```diff
@@ -37,16 +37,11 @@ class AWS0ClientTest extends AgentTestRunner {
     new ProfileCredentialsProvider(),
     new InstanceProfileCredentialsProvider())
 
-  def setupSpec() {
+  def setup() {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/test_before_1_11_106/groovy/AWS0ClientTest.groovy:40]: Why does it matter if this is `setup` or `setupSpec`?

File: dd-java-agent/instrumentation/jdbc/src/test/groovy/JDBCConnectionUrlParserTest.groovy [hunk 5]
```diff
@@ -1,10 +1,10 @@
+import datadog.trace.agent.test.AgentTestRunner
 import datadog.trace.bootstrap.instrumentation.jdbc.DBInfo
-import datadog.trace.test.util.DDSpecification
 import spock.lang.Shared
 
 import static datadog.trace.bootstrap.instrumentation.jdbc.JDBCConnectionUrlParser.parse
 
-class JDBCConnectionUrlParserTest extends DDSpecification {
+class JDBCConnectionUrlParserTest extends AgentTestRunner {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/jdbc/src/test/groovy/JDBCConnectionUrlParserTest.groovy:7]: What's the motivation for this change?

File: dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/test/groovy/RabbitMQTest.groovy [hunk 6]
```diff
@@ -47,12 +40,23 @@ class RabbitMQTest extends AgentTestRunner {
...
+    super.configurePreAgent()
+
+    injectSysConfig("dd.amqp.e2e.duration.enabled", "true")
+  }
+
+  def setup() {
+    factory = new ConnectionFactory(host: rabbitmqAddress.hostName, port: rabbitmqAddress.port)
+    conn = factory.newConnection()
+    channel = conn.createChannel()
+  }
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/test/groovy/RabbitMQTest.groovy:59]: Why did these need to be moved to `setup`?

File: dd-java-agent/src/test/groovy/datadog/trace/agent/JMXFetchTest.groovy [hunk 7]
```diff
@@ -80,7 +80,7 @@ class JMXFetchTest extends Specification {
 
     cleanup:
     names.each {
-      System.clearProperty("dd.integration.${it}.enabled")
+      System.clearProperty("dd.jmxfetch.${it}.enabled")
```
Comment (tylerbenson) [inline dd-java-agent/src/test/groovy/datadog/trace/agent/JMXFetchTest.groovy:83]: oops... Is this cleanup even needed anymore? I think you've generally removed all the other similar cleanup methods.

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.java [hunk 8]
```diff
@@ -285,7 +287,7 @@ public static void blockUntilChildSpansFinished(final int numberOfSpans) {
     }
   }
 
-  public static class TestRunnerListener implements AgentBuilder.Listener {
+  public static class TestRunnerListener extends AgentBuilder.Listener.Adapter {
```
Comment (tylerbenson) [inline dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.java:290]: 👍

File: dd-java-agent/testing/src/test/groovy/ConfigResetTest.groovy [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+  def setup() {
+    notSharedInstanceCreatedInSetup = new ShouldBeConstructedWithoutConfig()
+  }
+
+  static void checkStaticAssertions() {
+    assert System.getProperty("dd.trace.enabled") == null
+    assert System.getenv("DD_TRACE_ENABLED") == null
+    assert Config.get().isTraceEnabled()
+  }
```
Comment (tylerbenson) [inline dd-java-agent/testing/src/test/groovy/ConfigResetTest.groovy:29]: I think `ShouldBeConstructedWithoutConfig` is a bit confusing... consider calling this method directly and having this method return `new Object()` so it can be a valid assignment. (I don't think it matters that the assertion be done within the constructor.)

**Review by tylerbenson — COMMENTED** [review]
Nice cleanup.  I added a bunch of questions.

File: dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/test_before_1_11_106/groovy/AWS0ClientTest.groovy (hunk 4 above)
Comment (randomanderson) [reply dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/test_before_1_11_106/groovy/AWS0ClientTest.groovy:40]: System properties get reset to the pre-test stage before each test.  Properties set in `setupSpec()` will only apply to the very first test and get reset after.

File: dd-java-agent/instrumentation/jdbc/src/test/groovy/JDBCConnectionUrlParserTest.groovy (hunk 5 above)
Comment (randomanderson) [reply dd-java-agent/instrumentation/jdbc/src/test/groovy/JDBCConnectionUrlParserTest.groovy:7]: [This failure](https://app.circleci.com/pipelines/github/DataDog/dd-trace-java/4212/workflows/2c5e0ef1-3e30-4c5c-92c3-51c262b704a3/jobs/105197) .  There was that race where if this test class ran first, `Config` is loaded in the a application classloader instead of the the boot classloader

File: dd-java-agent/src/test/groovy/datadog/trace/agent/JMXFetchTest.groovy (hunk 7 above)
Comment (randomanderson) [reply dd-java-agent/src/test/groovy/datadog/trace/agent/JMXFetchTest.groovy:83]: `JMXFetchTest` doesn't extend `DDSpecification` (none of the top-level tests in `dd-java-agent` can).  Fixing this is part of the "future work" I described in the description.

File: dd-trace-core/src/test/groovy/datadog/trace/core/util/SystemAccessTest.groovy (hunk 2 above)
Comment (randomanderson) [reply dd-trace-core/src/test/groovy/datadog/trace/core/util/SystemAccessTest.groovy:96]: I considered that.  At first I had a mini DSL for setting config similar to `TagsAssert`.  After working through a bunch of the tests, I realized almost all tests that set a config only set one.

It just not worth it to complicate things.  (Even taking in a map involves the `GString` conversion step)

File: dd-trace-core/src/test/groovy/datadog/trace/core/CoreTracerTest.groovy [hunk 10]
```diff
@@ -31,22 +27,13 @@ import static datadog.trace.api.config.TracerConfig.PRIORITY_SAMPLING
...
 import static datadog.trace.api.config.TracerConfig.WRITER_TYPE
-import static datadog.trace.bootstrap.config.provider.SystemPropertiesConfigSource.PREFIX
 
 @Timeout(10)
 class CoreTracerTest extends DDSpecification {
 
-  @Rule
-  public final RestoreSystemProperties restoreSystemProperties = new RestoreSystemProperties()
-  @Rule
-  public final EnvironmentVariables environmentVariables = new EnvironmentVariables()
```
[the remark this answers is not in the document]
Comment (randomanderson) [reply dd-trace-core/src/test/groovy/datadog/trace/core/CoreTracerTest.groovy:42]: Yes. Done.

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/base/HttpServerTest.groovy (hunk 1 above)
Comment (randomanderson) [reply dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/base/HttpServerTest.groovy:318]: Consistency/style mostly.  It's cleaner to have "This is what's being tested" under `when` and "these are the expected results" under `then` so I usually just make the adjustment as I go.

There are some edge cases with mocks and others with `AsyncConditions` or `BlockingVariable` where putting things in the wrong block changes evaluation order.

File: dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/test/groovy/RabbitMQTest.groovy (hunk 6 above)
Comment (randomanderson) [reply dd-java-agent/instrumentation/rabbitmq-amqp-2.7/src/test/groovy/RabbitMQTest.groovy:59]: Instance fields get initialized before `setup()` methods are called.  This test set properties in a static block so when the fields get constructed, the properties are already set.  To maintain the same behavior, the field initialization had to be moved into the `setup()` block.

This test is still "inaccurate" because e2e duration in RabbitMQ is broken.  The bad config api hid that fact.  I listed it under "future work" to fix RabbitMQ.

**Review by tylerbenson — APPROVED** [review]

File: utils/test-utils/src/main/groovy/datadog/trace/test/util/DDSpecification.groovy [hunk 11]
```diff
@@ -43,6 +59,123 @@ abstract class DDSpecification extends Specification {
...
+    if (isConfigInstanceModifiable) {
+      rebuildConfig()
+    }
+  }
+
+  void setup() {
+    restoreProperties()
+
+    assert System.getenv().findAll { it.key.startsWith("DD_") }.isEmpty()
+    assert System.getProperties().findAll { it.key.toString().startsWith("dd.") }.isEmpty()
```
Comment (bantonsson) [inline utils/test-utils/src/main/groovy/datadog/trace/test/util/DDSpecification.groovy:110]: 👍 For the paranoid checks.

**Review by bantonsson — APPROVED** [review]
Great cleanup. Love that restoring is done _automatically_!
